bin/reflective_loader/src/loader.rs

```diff
@@ -0,0 +1,844 @@
+        if export_name_hash == dbj2_hash(name_slice) 
+        {
+            let ordinal = ordinals[i as usize] as usize;
+            return Some(module_base as usize + functions[ordinal] as usize);
+        }
+    }
+
+    return None;
+}
+
+/// Get the length of a C String
+pub unsafe fn get_cstr_len(pointer: *const char) -> usize 
+{
+    let mut tmp: u64 = pointer as u64;
+
+    while *(tmp as *const u8) != 0 
+    {
+        tmp += 1;
+    }
+
+    (tmp - pointer as u64) as _
+}
+
+#[no_mangle]
+pub fn reflective_loader(user_data_ptr_and_dll_bytes: *mut c_void) -> usize {
+    #[cfg(not(target_os = "windows"))]
+    panic!("This OS isn't supported by the dll_reflect function.\nOnly windows systems are supported");
+
+    // Dereference user data object
+    let user_data_ptr_ptr = user_data_ptr_and_dll_bytes as *const *const UserData;
+    let user_data_ptr = unsafe{ *user_data_ptr_ptr };
+    let user_data = unsafe{*user_data_ptr}.clone();
+    if user_data.function_offset == 0 { panic!("Could not parse the user_data segment") }
+    
+    // Increment dll_bytes ptr offset to after the user_data ptr.
+    let dll_bytes = (user_data_ptr_and_dll_bytes as usize + core::mem::size_of::<usize>()) as *mut c_void;
+
+    #[cfg(target_arch = "x86_64")]
+    let pe_header = PeFileHeaders64::new(dll_bytes);
+    if pe_header.dos_headers.e_magic != PE_MAGIC { panic!("Target DLL does not appear to be a DLL.") }
+
+    let kernel32_base = unsafe { get_loaded_module_by_hash(KERNEL32_HASH).unwrap() };
+    let ntdll_base = unsafe { get_loaded_module_by_hash(NTDLL_HASH).unwrap() };
+    if kernel32_base.is_null() || ntdll_base.is_null() 
+    {
+        panic!("Could not find kernel32 and ntdll");
+    }
+
+    // Create function pointers
+    // Get exports
+    let loadlib_addy = unsafe { get_export_by_hash(kernel32_base, LOAD_LIBRARY_A_HASH).expect("Couldn't lookup LoadLibraryA export by hash") };
+    let load_library_a_fn = unsafe { transmute::<_, FnLoadLibraryA>(loadlib_addy) };
+
+    let getproc_addy = unsafe { get_export_by_hash(kernel32_base, GET_PROC_ADDRESS_HASH).expect("Couldn't lookup GetProcAddress export by hash") };
+    let get_proc_address_fn = unsafe { transmute::<_, FnGetProcAddress>(getproc_addy) };
+
+    let virtualalloc_addy = unsafe { get_export_by_hash(kernel32_base, VIRTUAL_ALLOC_HASH).expect("Couldn't lookup VirtualAlloc export by hash") };
+    let virtual_alloc_fn = unsafe { transmute::<_, FnVirtualAlloc>(virtualalloc_addy) };
+
+    let getlasterror_addy = unsafe { get_export_by_hash(kernel32_base, GET_LAST_ERROR_HASH).expect("Couldn't lookup GetLastError export by hash") };
+    let get_last_error_fn = unsafe{ transmute::<_, FnGetLastError>(getlasterror_addy)};
+
+    // Allocate memory for our DLL to be loaded into
+    let new_dll_base: *mut c_void = virtual_alloc(virtual_alloc_fn, get_last_error_fn, ptr::null(), pe_header.nt_headers.OptionalHeader.SizeOfImage as usize, MEM_RESERVE | MEM_COMMIT, PAGE_EXECUTE_READWRITE);
```

Comment:
`MEM_COMMIT` on the whole DLL is a bit of a waste of space. It doesn't matter that much but it's better practice (and more like the windows loader) to `MEM_RESERVE` the whole region, then call `VirtualAlloc` again with `MEM_COMMIT` on all sections. Then once the DLL is loaded (before you call `DllMain` call `VirtualFree` on any sections with the `IMAGE_SCN_MEM_DISCARDABLE` bit set in `IMAGE_SECTION_HEADER.Characteristics`.
Also, `PAGE_EXECUTE_READWRITE` is bad practice. Apply section perms with `VirtualProtect`
You probably know both of those things already, but worth mentioning! 
How it is now is fine for a first version!

Reply:
Woot 🙌 Great call outs.
I think for now i'm gonna leave it as is.
I have a slight preference towards RWX and greedy allocation as it should help with getting detected.
Will definitely re-visit this advice when building the next one though. 🔥